services-core/src/solution_submission/retry.rs

```diff
@@ -10,170 +12,153 @@ use futures::{
+                args.batch_index,
+                args.solution.clone(),
+                args.claimed_objective_value,
                 num::f64_to_u256(gas_price),
-                nonce,
+                args.nonce,
             )
-            .map(FutureOutput::SolutionSubmission);
-        futures.push(solution_submission_future.boxed());
+            .map(FutureOutput::SolutionSubmission)
+            .boxed()
+    }
 
-        let is_last_iteration = gas_price >= effective_gas_price_cap;
-        if !is_last_iteration {
-            let timeout_future = async_sleep
-                .sleep(SINGLE_ATTEMPT_TIMEOUT)
-                .map(|()| FutureOutput::Timeout);
-            futures.push(timeout_future.boxed());
-        }
+    fn wait_interval(&self) -> BoxFuture<FutureOutput> {
+        self.async_sleep
+            .sleep(GAS_PRICE_REFRESH_INTERVAL)
+            .map(|()| FutureOutput::Timeout)
+            .boxed()
+    }
+
+    async fn retry_(&self, args: Args) -> Result<(), MethodError> {
+        let mut futures = FuturesUnordered::new();
+        let mut highest_used_gas_price = 0.0;
+        loop {
+            if let Some(gas_price) = self.new_gas_price(&args).await {
+                if gas_price >= highest_used_gas_price * MIN_GAS_PRICE_INCREASE_FACTOR {
+                    highest_used_gas_price = gas_price;
+                    futures.push(self.submit_solution(&args, gas_price));
+                }
+            }
 
-        // Like in `StableXSolutionSubmitter::submit_solution` we need to handle the situation where
-        // we observe a nonce error from one future before the completion of another. It is also
-        // possible that a previous submission transaction completes first instead of the most
-        // recent one.
-        // Unwrap because we always add the solution future above and every iteration here checks
-        // that there are still futures left.
-        while let FutureOutput::SolutionSubmission(result) = futures.next().await.unwrap() {
-            if !result.is_transaction_error() || futures.is_empty() {
-                return result;
+            futures.push(self.wait_interval());
+
+            // Like in `StableXSolutionSubmitter::submit_solution` we need to handle the situation where
+            // we observe a nonce error from one future before the completion of another. It is also
+            // possible that a previous submission transaction completes first instead of the most
+            // recent one.
+            // Unwrap because the timeout future always exists.
+            while let FutureOutput::SolutionSubmission(result) = futures.next().await.unwrap() {
+                // Comparison against `1` instead of `empty` because the timeout future always exists.
+                let is_last_transaction_future = futures.len() == 1;
+                if !result.is_transaction_error() || is_last_transaction_future {
+                    return result;
```

Comment:
Is this guaranteed to return? Why the change in logic from having the last iteration wait only for the solution submission future?

Reply:
In the previous version there was a clear last iteration when we reached the max gas price. At that point we wouldn't continue the `loop` anymore.
In this version we keep updating the gas price forever until a transaction completes or implicitly until the owner of this future drops it.
In both versions we keep track of previously sent transactions and have this while loop as the comment explains.